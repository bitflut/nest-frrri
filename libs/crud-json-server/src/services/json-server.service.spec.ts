import { Crud, CrudController, CrudDecoratorOptions, Endpoint } from '@frrri/nest-crud';
import { Controller, HttpModule } from '@nestjs/common';
import { NestApplication } from '@nestjs/core';
import { Test } from '@nestjs/testing';
import * as supertest from 'supertest';
import { JsonServerService, JsonServerServiceOptions } from './json-server.service';

interface Post {
    id: number;
    userId: number;
    body: string;
    title: string;
}

describe('JsonServerService', () => {

    describe('with query', () => {
        const collection = 'posts';
        const query = { id: [1, 2] } as CrudDecoratorOptions['query'];
        const endpoints = [Endpoint.GetMany, Endpoint.GetOne] as CrudDecoratorOptions['endpoints'];
        let app: NestApplication;
        let $: supertest.SuperTest<supertest.Test>;

        beforeAll(async () => {
            @JsonServerServiceOptions({ collection })
            class PostsService extends JsonServerService<Post> { }

            @Crud({ query, endpoints })
            @Controller({ path: 'posts' })
            class PostsController implements CrudController<Post> {
                constructor(public service: PostsService) { }
            }

            const moduleRef = await Test.createTestingModule({
                imports: [HttpModule],
                controllers: [PostsController],
                providers: [PostsService],
            }).compile();

            app = moduleRef.createNestApplication();
            await app.init();
            $ = supertest(app.getHttpServer());
        });

        it('should work', async () => {
            const getMany = await $.get('/posts').expect(200);
            expect(getMany.body).toMatchSnapshot('getMany(id=1&id=2)');
            expect(getMany.body.length).toEqual(2);
        });

        afterAll(async () => {
            await app.close();
        });
    });

    describe('with pagination', () => {
        const collection = 'posts';
        const query = { id: [1, 2, 3, 4, 5, 6], _limit: 3 } as CrudDecoratorOptions['query'];
        const endpoints = [Endpoint.GetMany, Endpoint.GetOne] as CrudDecoratorOptions['endpoints'];
        let app: NestApplication;
        let $: supertest.SuperTest<supertest.Test>;

        beforeAll(async () => {
            @JsonServerServiceOptions({ collection })
            class PostsService extends JsonServerService<Post> { }

            @Crud({ query, endpoints })
            @Controller({ path: 'posts' })
            class PostsController implements CrudController<Post> {
                constructor(public service: PostsService) { }
            }

            const moduleRef = await Test.createTestingModule({
                imports: [HttpModule],
                controllers: [PostsController],
                providers: [PostsService],
            }).compile();

            app = moduleRef.createNestApplication();
            app.enableCors({
                allowedHeaders: ['link'],
                exposedHeaders: ['link'],
            });
            await app.init();
            $ = supertest(app.getHttpServer());
        });

        it('should work', async () => {
            const getMany = await $.get('/posts?_page=1').expect(200);
            expect(getMany.body).toMatchSnapshot('getMany(pagination)');
            expect(getMany.body.length).toEqual(3);

            const linkHeader: string = getMany.header['link'];
            expect(linkHeader).toBeDefined();

            const nextRegex = /<([^>]+)>;\s?rel="([^"]+)"/;
            const links: any = linkHeader.match(new RegExp(nextRegex, 'g'))
                .reduce((prev, curr) => {
                    const res = curr.match(new RegExp(nextRegex));
                    prev[res[2]] = res[1].replace(/.+\/([^\/]+)$/, '/$1');
                    return prev;
                }, {});
            expect(links.next).toBeDefined();
            expect(links.last).toBeDefined();

            const getNext = await $.get(links.next).expect(200);
            expect(getNext.body).toMatchSnapshot('getNext(pagination)');
            expect(getNext.body.length).toEqual(3);
        });

        afterAll(async () => {
            await app.close();
        });
    });

});
