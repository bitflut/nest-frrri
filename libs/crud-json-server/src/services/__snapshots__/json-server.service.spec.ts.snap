// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonServerService with pagination should work: getMany(pagination) 1`] = `
Array [
  Object {
    "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "userId": 1,
  },
  Object {
    "body": "est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla",
    "id": 2,
    "title": "qui est esse",
    "userId": 1,
  },
  Object {
    "body": "et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut",
    "id": 3,
    "title": "ea molestias quasi exercitationem repellat qui ipsa sit aut",
    "userId": 1,
  },
]
`;

exports[`JsonServerService with pagination should work: getNext(pagination) 1`] = `
Array [
  Object {
    "body": "ullam et saepe reiciendis voluptatem adipisci
sit amet autem assumenda provident rerum culpa
quis hic commodi nesciunt rem tenetur doloremque ipsam iure
quis sunt voluptatem rerum illo velit",
    "id": 4,
    "title": "eum et est occaecati",
    "userId": 1,
  },
  Object {
    "body": "repudiandae veniam quaerat sunt sed
alias aut fugiat sit autem sed est
voluptatem omnis possimus esse voluptatibus quis
est aut tenetur dolor neque",
    "id": 5,
    "title": "nesciunt quas odio",
    "userId": 1,
  },
  Object {
    "body": "ut aspernatur corporis harum nihil quis provident sequi
mollitia nobis aliquid molestiae
perspiciatis et ea nemo ab reprehenderit accusantium quas
voluptate dolores velit et doloremque molestiae",
    "id": 6,
    "title": "dolorem eum magni eos aperiam quia",
    "userId": 1,
  },
]
`;

exports[`JsonServerService with query should work: getMany(id=1&id=2) 1`] = `
Array [
  Object {
    "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "userId": 1,
  },
  Object {
    "body": "est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla",
    "id": 2,
    "title": "qui est esse",
    "userId": 1,
  },
]
`;
